src/EditorFeatures/Core/Implementation/Formatting/FormatCommandHandler.FormatDocument.cs

```diff
@@ -34,18 +60,32 @@ private bool TryExecuteCommand(FormatDocumentCommandArgs args, CommandExecutionC
                 return false;
             }
 
-            var formattingService = document.GetLanguageService<IEditorFormattingService>();
-            if (formattingService == null || !formattingService.SupportsFormatDocument)
+            if (!document.Project.Solution.Workspace.Options.GetOption(FeatureOnOffOptions.IsCodeCleanupRulesConfigured, LanguageNames.CSharp))
             {
-                return false;
+                ShowGoldBarForCodeCleanupConfiguration(document.Project.Solution.Workspace);
             }
 
             using (context.WaitContext.AddScope(allowCancellation: true, EditorFeaturesResources.Formatting_document))
             {
-                Format(args.TextView, document, null, context.WaitContext.UserCancellationToken);
-            }
+                var cancellationToken = context.WaitContext.UserCancellationToken;
+
+                using (var transaction = new CaretPreservingEditTransaction(
+                    EditorFeaturesResources.Formatting, args.TextView, _undoHistoryRegistry, _editorOperationsFactoryService))
+                {
+                    document = _codeCleanupService.CleanupDocument(document, cancellationToken);
```

Comment:
this should be the one that apply change to worksapce. I thought we said we will have 1 undo transaction?

Reply:
updated.